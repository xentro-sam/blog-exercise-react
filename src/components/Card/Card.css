.card {
    width: 25vw;
    height: max-content;
    border: 1px solid grey;
    margin-top: 1%;
    margin-bottom: 1%;
}

.card-image > img {
    width: 100%;
    height: 45vh;
}

.card-data {
    display: flex;
    justify-content: space-between;
    color: gray;
}

.card-data > div {
    padding: 10px;
}

.card-title {
    font-size: 1.5rem;
    padding: 10px;
    min-height: 7vh;
}

.card-subtitle {
    font-size: 1rem;
    color: gray;
    padding: 10px;
}

.card-footer {
    display: flex;
    justify-content: space-between;
}

.card-claps {
    display: flex;
    color: gray;
}

.claps {
    padding-top: 10px;
}

.card-footer-icons > img {
    width: 20px;
    height: 20px;
    padding: 10px;
}
