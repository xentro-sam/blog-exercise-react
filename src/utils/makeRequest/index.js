import axios from 'axios';
import {BACKEND_URL} from '../../constants/apiEndPoints';
import {ERROR_ROUTE} from '../../constants/routes';

const makeRequest = async (apiEndPoint, dynamicConfig = {}, navigate) => {
  try {
    const requestDetails = {
      baseURL: `${BACKEND_URL}${apiEndPoint.url}`,
      method: apiEndPoint.method,
      ...dynamicConfig,
    };
    const {data} = await axios(requestDetails);
    return data;
  } catch (error) {
    if (navigate) {
      const errorStatus = error.response?.status;
      if (errorStatus) {
        navigate(`${ERROR_ROUTE}/${errorStatus}`);
      } else {
        navigate(ERROR_ROUTE);
      }
    }
  }
};

export default makeRequest;
