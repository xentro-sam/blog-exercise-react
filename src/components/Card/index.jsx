import * as React from 'react';
import './Card.css';

import clappingIcon from '../../assets/Icons/clapping.svg';
import likeIcon from '../../assets/Icons/heart-red.svg';
import unlikeIcon from '../../assets/Icons/heart-black.svg';

export default function Card(cardInfo) {
  const [like, setLike] = React.useState(false);
  const [claps, setClaps] = React.useState(cardInfo.claps);
  const imageSrc = require(`../../assets/Images/${cardInfo.image}`);
  return (
    <div className="card">
      <div className="card-image">
        <img src={imageSrc} alt="" />
      </div>
      <div className="card-data">
        <div>{cardInfo.date}</div>
        <div>{cardInfo.readingTime}</div>
      </div>
      <div className="card-title">
        {cardInfo.title}
      </div>
      <div className="card-subtitle">
        {cardInfo.description}
      </div>
      <div>
        <hr />
      </div>
      <div className="card-footer">
        <div className="card-claps">
          <div className="card-footer-icons">
            <img src={clappingIcon} alt="" onClick={() => setClaps(claps + 1)} />
          </div>
          <div className="claps">
            {claps}
          </div>
        </div>
        <div className="card-footer-icons">
          <img
            src={like ? likeIcon : unlikeIcon}
            alt=""
            onClick={() => setLike(!like)}
          />
        </div>
      </div>
    </div>
  );
}
