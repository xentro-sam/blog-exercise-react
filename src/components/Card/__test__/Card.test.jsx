import React from 'react';
import {fireEvent, render, waitFor} from '@testing-library/react';
import Card from '..';
import makeRequest from '../../../utils/makeRequest';

jest.mock('../../../utils/makeRequest');

jest.mock('react-router-dom', () => ({
  useNavigate: () => jest.fn(),
}));

const cardInfo = {
  id: '1',
  date: '2020-10-10T00:00:00.000Z',
  reading_time: '5 mins',
  title: 'Test Title',
  description: 'Test Description',
  image: 'https://picsum.photos/200/300',
  liked: false,
  claps: 0,
};

describe('Card', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it('should render correctly', () => {
    makeRequest.mockResolvedValueOnce({data: {claps: 1}});
    const {asFragment} = render(<Card key={cardInfo.id} {...cardInfo} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should increment claps', async () => {
    const {getByAltText, getByTestId} = render(<Card key={cardInfo.id} {...cardInfo} />);
    const clappingIcon = getByAltText('clapping');
    const clapCount = getByTestId('clap-count');
    expect(clapCount.textContent).toBe('0');
    makeRequest.mockResolvedValueOnce({data: {claps: 1}});
    expect(makeRequest).not.toHaveBeenCalled();
    fireEvent.click(clappingIcon);
    await waitFor(() => {
      expect(makeRequest).toHaveBeenCalled();
      expect(clapCount.textContent).toBe('1');
    });
  });
  it('should toggle like', async () => {
    const {getByAltText} = render(<Card key={cardInfo.id} {...cardInfo} />);
    const likeIcon = getByAltText('like');
    expect(likeIcon.src).toContain('heart-black.svg');
    makeRequest.mockResolvedValueOnce({data: {liked: true}});
    expect(makeRequest).not.toHaveBeenCalled();
    fireEvent.click(likeIcon);
    await waitFor(() => {
      expect(makeRequest).toHaveBeenCalled();
      expect(likeIcon.src).toContain('heart-red.svg');
    });
  });
});
