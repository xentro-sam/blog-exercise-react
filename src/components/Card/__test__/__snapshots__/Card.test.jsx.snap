// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should render correctly 1`] = `
<DocumentFragment>
  <div
    class="card"
  >
    <div
      class="card-image"
    >
      <img
        alt="image"
        src="https://picsum.photos/200/300"
      />
    </div>
    <div
      class="card-data"
    >
      <div>
        10th
      October, 
      2020
      </div>
      <div>
        5 mins
      </div>
    </div>
    <div
      class="card-title"
    >
      Test Title
    </div>
    <div
      class="card-subtitle"
    >
      Test Description
    </div>
    <div>
      <hr />
    </div>
    <div
      class="card-footer"
    >
      <div
        class="card-claps"
      >
        <div
          class="card-footer-icons"
        >
          <img
            alt="clapping"
            src="clapping.svg"
          />
        </div>
        <div
          class="claps"
          data-testid="clap-count"
        >
          0
        </div>
      </div>
      <div
        class="card-footer-icons"
      >
        <img
          alt="like"
          src="heart-black.svg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
